modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java

```diff
@@ -0,0 +1,365 @@
+        // FileAlreadyExistsException is thrown and this method silently returns.
+        // (this method is never invoked concurrently and is invoked by a cluster state applier thread)
+        final Path databaseTmpGzFile;
+        try {
+            databaseTmpGzFile = Files.createFile(geoipTmpDirectory.resolve(databaseName + ".tmp.gz"));
+        } catch (FileAlreadyExistsException e) {
+            LOGGER.debug("database update [{}] already in progress, skipping...", databaseName);
+            return;
+        }
+
+        final Path databaseTmpFile = Files.createFile(geoipTmpDirectory.resolve(databaseName + ".tmp"));
+        LOGGER.info("downloading geoip database [{}] to [{}]", databaseName, databaseTmpGzFile);
+        retrieveDatabase(
+            databaseName,
+            recordedMd5,
+            metadata,
+            bytes -> Files.write(databaseTmpGzFile, bytes, StandardOpenOption.APPEND),
+            () -> {
+                LOGGER.debug("decompressing [{}]", databaseTmpGzFile.getFileName());
+                decompress(databaseTmpGzFile, databaseTmpFile);
+
+                Path databaseFile = geoipTmpDirectory.resolve(databaseName);
+                LOGGER.debug("moving database from [{}] to [{}]", databaseTmpFile, databaseFile);
+                Files.move(databaseTmpFile, databaseFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+                updateDatabase(databaseName, recordedMd5, databaseFile);
+                Files.delete(databaseTmpGzFile);
+            },
+            failure -> {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to download database [{}]", databaseName), failure);
+                try {
+                    Files.delete(databaseTmpFile);
+                    Files.delete(databaseTmpGzFile);
+                } catch (IOException ioe) {
+                    ioe.addSuppressed(failure);
+                    LOGGER.error("Unable to delete tmp database file after failure", ioe);
+                }
+            });
+    }
+
+    void updateDatabase(String databaseFileName, String recordedMd5, Path file) {
+        try {
+            LOGGER.info("database file changed [{}], reload database...", file);
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file);
+            DatabaseReference existing = databases.put(databaseFileName, new DatabaseReference(recordedMd5, file, loader));
+            if (existing != null) {
+                existing.close();
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    void removeStaleEntries(Collection<String> staleEntries) {
+        try {
+            List<DatabaseReference> references = new ArrayList<>();
```

Comment:
do we need this list? can we just merge these two `for` loops?

Reply:
yeah i think merging is possible

Reply:
I think this is a direct copy from the poc, which did something silly.